glass/src/lib/native/cpp/support/ExpressionParser.cpp

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "glass/support/ExpressionParser.h"
+
+#include <cctype>
+#include <cmath>
+#include <stack>
+#include <string>
+
+namespace glass::expression {
+
+template <typename V>
+ParseResult<V> ParseResult<V>::CreateSuccess(V value) {
+  ParseResult<V> result;
+  result.kind = ParseResultKind::Success;
+  result.successVal = value;
+  return result;
+}
+
+template <typename V>
+ParseResult<V> ParseResult<V>::CreateError(const char* errorMessage) {
+  ParseResult<V> result;
+  result.kind = ParseResultKind::Error;
+  result.errorMessage = errorMessage;
+  return result;
+}
+
+enum class TokenType {
+  Number,
+  Add,
+  Subtract,
+  Multiply,
+  Divide,
+  Exponent,
+  OpenParen,
+  CloseParen,
+
+  End,    // Hit end of input
+  Error,  // Invalid character found
+};
+
+struct Token {
+  TokenType type;
+  const char* str;
+  int strLen;
+
+  explicit Token(TokenType type) : type(type), str(nullptr), strLen(0) {}
+
+  Token(TokenType type, const char* str, int strLen)
+      : type(type), str(str), strLen(strLen) {}
+};
+
+class Lexer {
+ public:
+  explicit Lexer(const char* input) : input(input) {}
+
+  Token NextToken() {
+    // Skip leading whitespace
+    startIdx = currentIdx;
+    while (std::isspace(input[startIdx])) {
+      startIdx++;
+    }
+    if (input[startIdx] == 0) {
+      return Token(TokenType::End);
+    }
+    currentIdx = startIdx;
+
+    char c = input[currentIdx];
+    currentIdx++;
+    switch (c) {
+      case '+':
+        return Token(TokenType::Add);
+      case '-':
+        return Token(TokenType::Subtract);
+      case '*':
+        return Token(TokenType::Multiply);
+      case '/':
+        return Token(TokenType::Divide);
+      case '^':
+        return Token(TokenType::Exponent);
+      case '(':
+        return Token(TokenType::OpenParen);
+      case ')':
+        return Token(TokenType::CloseParen);
+      default:
+        currentIdx--;
+        if (std::isdigit(c) || c == '.') {
+          return nextNumber();
+        }
+        return Token(TokenType::Error);
+    }
+  }
+
+  // Makes NextToken() return the same token as previously
+  void Repeat() { currentIdx = startIdx; }
+
+ private:
+  const char* input;
+  int startIdx = 0, currentIdx = 0;
+
+  Token nextNumber() {
+    // Read whole part
+    bool hasDigitsBeforeDecimal = false;
+    while (std::isdigit(input[currentIdx])) {
+      currentIdx++;
+      hasDigitsBeforeDecimal = true;
+    }
+
+    // Read decimal part if it exists
+    if (input[currentIdx] == '.') {
+      currentIdx++;
+      if (!hasDigitsBeforeDecimal && !std::isdigit(input[currentIdx])) {
```

Comment:
`currentIdx` is incremented on the line above, so that is checking if the character after the `'.'` is a digit. Essentially this part is making sure the number actually has any digits and isn't just a decimal point.

Reply:
Whoops, I missed that!